nexus/src/app/background/tasks/support_bundle_collector.rs

```diff
@@ -722,152 +806,322 @@ impl BundleCollection {
+        mgs_client
+            .get_or_init(|| async {
+                Arc::new(self.create_mgs_client().await.ok())
+            })
+            .await
+    }
 
-            const MAX_CONCURRENT_SLED_REQUESTS: usize = 16;
-            const FAILURE_MESSAGE: &str =
-                "Failed to fully collect support bundle info from sled";
-            let mut set = ParallelTaskSet::new_with_parallelism(
-                MAX_CONCURRENT_SLED_REQUESTS,
+    async fn get_or_initialize_all_sleds<'a>(
+        &self,
+        all_sleds: &'a OnceCell<Arc<Option<Vec<Sled>>>>,
+    ) -> &'a Arc<Option<Vec<Sled>>> {
+        all_sleds
+            .get_or_init(|| async {
+                Arc::new(
+                    self.datastore
+                        .sled_list_all_batched(
+                            &self.opctx,
+                            SledFilter::InService,
+                        )
+                        .await
+                        .ok(),
+                )
+            })
+            .await
+    }
+
+    async fn collect_sled_cubby_info(
+        &self,
+        all_sleds: &OnceCell<Arc<Option<Vec<Sled>>>>,
+        mgs_client: &OnceCell<Arc<Option<MgsClient>>>,
+        dir: &Utf8Path,
+    ) -> anyhow::Result<CollectionStepOutput> {
+        let Some(mgs_client) =
+            &**self.get_or_initialize_mgs_client(mgs_client).await
+        else {
+            warn!(
+                self.log,
+                "No MGS client, skipping sled cubby info collection"
             );
+            return Ok(CollectionStepOutput::None);
+        };
+        let nexus_sleds = self
+            .get_or_initialize_all_sleds(all_sleds)
+            .await
+            .as_deref()
+            .unwrap_or_default();
+
+        write_sled_cubby_info(&self.log, mgs_client, nexus_sleds, dir).await?;
+
+        Ok(CollectionStepOutput::None)
+    }
+
+    async fn spawn_sp_dump_collection(
+        &self,
+        mgs_client: &OnceCell<Arc<Option<MgsClient>>>,
+        dir: &Utf8Path,
+    ) -> anyhow::Result<CollectionStepOutput> {
+        let Some(mgs_client) =
+            &**self.get_or_initialize_mgs_client(mgs_client).await
+        else {
+            warn!(self.log, "No MGS client, skipping SP task dump collection");
+            return Ok(CollectionStepOutput::None);
+        };
 
-            for sled in all_sleds {
-                let prev_result = set
-                    .spawn({
-                        let collection: Arc<BundleCollection> = self.clone();
-                        let dir = dir.path().to_path_buf();
+        let sp_dumps_dir = dir.join("sp_task_dumps");
+        tokio::fs::create_dir_all(&sp_dumps_dir).await.with_context(|| {
+            format!("Failed to create SP task dump directory {sp_dumps_dir}")
+        })?;
+
+        let mut extra_steps: Vec<(&'static str, CollectionStepFn)> = vec![];
+        for sp in get_available_sps(&mgs_client).await? {
+            extra_steps.push((
```

Comment:
Rather than collecting all SP reports via new tokio tasks, these are creating new "steps" that get kicked back to the top-level `ParallelTaskSet` I mentioned earlier.